build: link libluv / luv.so by full path

blueyed opened the pull request:
This is meant to help with linking against "luv.so", which does not work
with "-lluv", if "luv.so" should be used.

Ref: https://github.com/NixOS/nixpkgs/issues/64400#issuecomment-516557253

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -401,10 +401,16 @@ endif()
   cmake_policy(SET CMP0069 NEW)
 endif()
 
+# Link libraries by full path even in implicit directories.
```
```suggestion
# Link libraries by full path even in implicit directories. #10407
```

blueyed committed:
cmake: libluv: look for luv.so, work around -lluv

Use "luv" as imported library to work around "-lluv" being used due to
missing SONAME.

Fixes https://github.com/neovim/neovim/issues/10407.
Ref: https://gitlab.kitware.com/cmake/cmake/issues/19578
Ref: https://github.com/NixOS/nixpkgs/issues/64400#issuecomment-516557253

blueyed commented:
Backed out 4f2c04ff7 - adding `luv.so` to NAMES fixes it.
I've tested it by having `luv.so` myself only.

blueyed commented:
Waiting for confirmation via https://github.com/NixOS/nixpkgs/issues/64400 / #10407.

blueyed commented on cmake/FindLibLUV.cmake:
```diff
@@ -20,7 +20,8 @@ if(LIBLUV_USE_STATIC)
 
-list(APPEND LIBLUV_NAMES luv)
+# Explicitly look for luv.so. #10407
+list(APPEND LIBLUV_NAMES luv luv${CMAKE_SHARED_LIBRARY_SUFFIX})
```
I think this might make sense in general, but does not help by itself usually (at least for nixpkgs).
A better approach is passing linker flags directly: https://github.com/NixOS/nixpkgs/issues/64400#issuecomment-517930779

justinmk replied:
Should at least mention that in https://github.com/neovim/neovim/wiki/Building-Neovim , or maybe in the main README.md

blueyed commented on src/nvim/CMakeLists.txt:
```diff
@@ -401,10 +401,16 @@ endif()
+# Link libraries by full path even in implicit directories. #10407
+if(POLICY CMP0060)
+  cmake_policy(SET CMP0060 NEW)
+endif()
```
This appears to make no difference really - or at least is not enough by itself.

mcepl commented:
yes, as @blueyed in https://github.com/neovim/neovim/issues/10407#issuecomment-517942811 said, this really doesn’t seem to help. ``-DLIBLUV_LIBRARY:STRING=-Wl,/usr/lib64/lua/5.1/luv.so`` did help, but it is even more ugly, than one would believe possible.

blueyed commented:
> but it is even more ugly, than one would believe possible.

Is it?
It is not much different than specifying the full path in the first place - it's just that you have to know about the possibility to pass in linker flags directly.

mcepl commented:
> It is not much different than specifying the full path in the first place - it's just that you have to know about the possibility to pass in linker flags directly.

Well, the point of all those CMake*.txt files is to have *declarative* description of the program needs. Here we have one hack (using gcc option instead of declaration of the library name as expected) on the top of another hack (pushing linker parameters through gcc).

blueyed commented:
Yes, it is unfortunate that `CMP0060` does not enable it like I've expected it to do.

With `cmake build -D LIBLUV_LIBRARY=/home/daniel/Vcs/luv/luv.so -DLIBLUV_INCLUDE_DIR=/home/daniel/Vcs/luv/include` it uses `-Wl,-rpath,/home/daniel/Vcs/luv -lluv` regardless of the policy, but e.g. `/usr/lib/libuv.so` instead of `-luv`..!

The problem is that `luv.so` is missing `SONAME`.
(As a workaround `set(CMAKE_PLATFORM_USES_PATH_WHEN_NO_SONAME 0)` works, but it is not considered to be safe.)
I've filed https://gitlab.kitware.com/cmake/cmake/issues/19578 - a workaround is mentioned there that I will use here.

blueyed commented:
Fails on AppVeyor:
```
-- Found LibLUV: C:/projects/nvim-deps/usr/lib/libluv.a (Required is at least version "1.30.0") 
…
-- Build files have been written to: C:/projects/neovim/build
DEBUG:  109+ cmake -G $cmakeGenerator $(convertToCmakeArgs($nvimCmakeVars)) .. ;  >>>> exitIfFailed
DEBUG:   24+ function exitIfFailed()  >>>> {
DEBUG:   25+   if ( >>>> $LastExitCode -ne 0) {
DEBUG:   29+  >>>> }
DEBUG:  110+  >>>> cmake --build . --config $cmakeBuildType -- $cmakeGeneratorArgs ; exitIfFailed
ninja: error: 'luv-NOTFOUND', needed by 'bin/nvim.exe', missing and no known rule to make it
```

blueyed commented:
@mcepl 
Does this approach here now work for you?

mcepl commented:
Yes, it not a solution, but certainly a good workaround. You can close this, this PR is mostly useless, IMHO.

blueyed commented:
@mcepl 
This PR is meant to fix it, via a workaround.

It allows to specify `-DLIBLUV_LIBRARY=/path/to/luv/luv.so` - without linker flags etc.
(And it should also enable finding `luv.so` in standard paths (which is not the case with lua dirs though probably).)

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -401,6 +406,7 @@ endif()
 
 set(NVIM_EXEC_LINK_LIBRARIES ${NVIM_LINK_LIBRARIES} ${LUA_PREFERRED_LIBRARIES})
 
+# Enable LTO (Link Time Optimization) #8654
```
```suggestion
# For LTO (Link Time Optimization) #8654
```

https://cmake.org/cmake/help/latest/policy/CMP0069.html

Alternatively:

```suggestion
# Add IPO flags (for LTO), or error if CMake does not know the flags. #8654
```

blueyed committed:
Update src/nvim/CMakeLists.txt  [ci skip]

Co-Authored-By: Justin M. Keyes <justinkz@gmail.com>

mcepl commented:
> @mcepl
> It allows to specify `-DLIBLUV_LIBRARY=/path/to/luv/luv.so` - without linker flags etc.
> (And it should also enable finding `luv.so` in standard paths (which is not the case with lua dirs though probably).)

Right, thank you. Yes, I can confirm, that with the latest version of this PR, I can use cmake option ``-DLIBLUV_LIBRARY=/path/to/luv/luv.so`` and it works. Thank you, very much.

blueyed commented:
\o/
